File: scripts/stop_watchdog.py
# scripts/stop_watchdog.py
"""
5-minute stop-loss watchdog (roadmap 2026-07-10 item 7).

Cut losses fast: the hourly evaluation is the only place stops were checked,
so a move between evaluations could run unprotected for up to an hour (and
historically much longer — see Bug #4). This script checks ONLY the exit
levels of open long positions and fires the existing trade.py stop path.
No research, no scoring, no new entries.

Checks per open long, in priority order (same rules as run_evaluation):
  1. Pending SELL order for the symbol -> skip (deduplication; covers stops
     placed by the hourly engine, the dashboard Autopilot, or a previous
     watchdog run).
  2. Trailing stop: HWM from data/positions_state.json; trail width fixed or
     Chandelier (risk.trail_mode) — identical to the hourly engine.
  3. Hard stop: max(4H swing low, breakeven stop) with the fixed -5% fallback.

All orders go through trade.py (limit-only, clamped stop band). State writes
are limited to recording the placed stop order (set_stop_order) so the hourly
engine's dedup/escalation sees it.

Usage:
    python scripts/stop_watchdog.py            # dry-run
    python scripts/stop_watchdog.py --execute  # place stop orders
"""

from __future__ import annotations

import argparse
import sys
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo

import _env  # noqa: F401  -- load .env into os.environ
import indicators as ind
import position_state as ps
from symbols import to_slash
from risk import (
    STOP_LOSS_MODE,
    STOP_LOSS_PCT,
    TRAIL_MODE,
    chandelier_trail_pct,
    should_stop_out,
    should_trail_stop_out,
    stop_loss_limit_price,
    swing_low_stop_price,
)
from trade import (
    TradeRejected,
    get_latest_quote,
    get_open_orders,
    get_positions,
    place_order,
)

PROJECT_ROOT = Path(__file__).resolve().parent.parent
JOURNAL_DIR  = PROJECT_ROOT / "journal"


def check_position(pos: dict, state: dict, execute: bool) -> str | None:
    """Check one open long position; returns a journal line when it acted."""
    sym  = to_slash(pos.get("symbol", ""))
    qty  = float(pos.get("qty") or 0)
    if qty <= 0:
        return None  # long-only

    # Dedup: any pending SELL for this symbol means an exit is in flight.
    try:
        if any(o.get("side") == "sell" for o in get_open_orders(sym)):
            return None
    except Exception:
        return None  # can't verify -> fail safe, do nothing

    ps_pos = ps.get_position(state, sym)
    entry  = float(pos.get("avg_entry_price") or 0)
    if entry <= 0:
        entry = float(ps_pos.get("entry_price") or 0)
    if entry <= 0:
        return None  # no usable cost basis — leave it to the hourly engine

    try:
        q   = get_latest_quote(sym)
        ask = float(q.get("ap") or 0)
        bid = float(q.get("bp") or 0)
    except Exception:
        return None
    cur = bid or ask
    if cur <= 0 or ask <= 0:
        return None

    # 4H bars for the swing-low stop (and ATR for the chandelier trail).
    swing_stop = None
    atr_4h     = None
    if STOP_LOSS_MODE == "swing_low_4h":
        try:
            from run_evaluation import get_crypto_bars_4h
            bars_4h = get_crypto_bars_4h(sym)
            lows    = [float(b.get("l") or 0) for b in bars_4h if b.get("c")]
            highs   = [float(b.get("h") or 0) for b in bars_4h if b.get("c")]
            closes  = [float(b.get("c") or 0) for b in bars_4h if b.get("c")]
            swing_stop = swing_low_stop_price(entry, lows)
            if len(closes) >= 15:
                atr_4h = ind.atr(highs, lows, closes)
        except Exception:
            pass  # fixed -5% fallback below

    # Trailing stop first (supersedes the hard stop once armed).
    hwm = ps_pos.get("high_water_mark") or entry
    trail_pct = None
    if TRAIL_MODE == "chandelier":
        trail_pct = chandelier_trail_pct(cur, atr_4h)
    trail_hit = (should_trail_stop_out(entry, hwm, cur, trail_pct=trail_pct)
                 if trail_pct is not None
                 else should_trail_stop_out(entry, hwm, cur))

    breakeven = ps_pos.get("breakeven_stop")
    eff_stop  = max([s for s in (swing_stop, breakeven) if s], default=None)
    hard_hit  = should_stop_out(entry, cur, stop_price=eff_stop)

    if not (trail_hit or hard_hit):
        return None

    kind = "TRAILING STOP" if trail_hit else (
        "STOP (breakeven)" if eff_stop and breakeven and eff_stop == breakeven
        else "STOP (4H swing low)" if eff_stop
        else "STOP (fixed -%d%%)" % int(STOP_LOSS_PCT * 100))
    lim = stop_loss_limit_price(ask, ps_pos.get("stop_order_cycles", 0))
    line = ("%s %s: entry $%.4f current $%.4f -> SELL %.4f @ $%.4f"
            % (sym, kind, entry, cur, qty, lim))

    if not execute:
        return line + " (dry-run)"
    try:
        result = place_order(sym, qty, "sell", lim, is_stop_loss=True)
        order_id = result.get("id", "")
        if order_id:
            ps.set_stop_order(state, sym, order_id, lim)
        return line + " id=" + order_id[:8]
    except TradeRejected as e:
        return line + " REJECTED: " + str(e)
    except Exception as e:
        return line + " ERROR: " + repr(e)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--execute", action="store_true",
                        help="place stop orders (default is dry-run)")
    args = parser.parse_args()

    state = ps.load_state()
    try:
        positions = get_positions()
    except Exception as e:
        sys.stderr.write("FAIL: positions fetch: %r\n" % e)
        return 1

    longs = [p for p in positions if float(p.get("qty") or 0) > 0]
    actions = []
    for p in longs:
        line = check_position(p, state, args.execute)
        if line:
            actions.append(line)
            print(line)

    if not actions:
        print("stop watchdog: %d long position(s) checked, no stops hit"
              % len(longs))
        return 0

    # Journal + state only when something happened (a no-op every 5 minutes
    # must not generate commits).
    if args.execute:
        ps.save_state(state)
    now = datetime.now(ZoneInfo("Europe/Amsterdam"))
    JOURNAL_DIR.mkdir(exist_ok=True)
    path = JOURNAL_DIR / (now.strftime("%Y-%m-%d") + ".md")
    with path.open("a", encoding="utf-8") as f:
        f.write("\n## Stop Watchdog %s GMT+2\n\n" % now.strftime("%H:%M"))
        for line in actions:
            f.write("- " + line + "\n")
    print("Wrote: " + str(path))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/stopWatchdog.js
// src/stopWatchdog.js
//
// Stop-loss watchdog (roadmap 2026-07-10 item 7) — a faithful port of
// scripts/stop_watchdog.py. Checks ONLY the exit levels of open long
// positions and fires the existing trade.js stop path. No research, no
// scoring, no new entries — decoupled from the hourly evaluation cycle so a
// move between evaluations doesn't run unprotected.
//
// Checks per open long, in priority order (identical to run_evaluation):
//   1. Pending SELL order for the symbol -> skip (dedup; covers stops placed
//      by the hourly engine, the dashboard Autopilot, or a previous
//      watchdog run).
//   2. Trailing stop: HWM from positions_state.json.
//   3. Hard stop: max(4H swing low, breakeven) with the fixed -5% fallback.
//
// Every external effect is injectable via `deps` (same pattern as
// runEvaluation.js) so the decision logic is unit-testable without HTTP
// stubbing.

import { pathToFileURL } from "node:url";
import { appendFileSync, mkdirSync } from "node:fs";
import path from "node:path";
import * as ps from "./positionState.js";
import { toSlash } from "./symbols.js";
import { amsterdamParts } from "./tz.js";
import {
  getPositions as defaultGetPositions,
  getOpenOrders as defaultGetOpenOrders,
  getLatestQuote as defaultGetLatestQuote,
  placeOrder as defaultPlaceOrder,
  TradeRejected,
} from "./trade.js";
import { getCryptoBars4h as defaultGetCryptoBars4h } from "./marketData.js";
import { STOP_LOSS_MODE, STOP_LOSS_PCT, TRAIL_MODE, shouldStopOut, shouldTrailStopOut, stopLossLimitPrice, swingLowStopPrice } from "./risk.js";
import { assertNotShipped } from "./strategyConfig.js";
import { JOURNAL_DIR } from "./journal.js";

// Chandelier trailing is not ported to risk.js (see risk.js's scope note) —
// fail loudly rather than silently falling back to the fixed trail if it's
// ever switched on before that lands.
assertNotShipped("risk.trail_mode", TRAIL_MODE === "chandelier", "chandelierTrailPct");

/** Check one open long position. Returns a journal line when it acted, else null. */
export async function checkPosition(pos, state, execute, deps = {}) {
  const getOpenOrders = deps.getOpenOrders || defaultGetOpenOrders;
  const getLatestQuote = deps.getLatestQuote || defaultGetLatestQuote;
  const getCryptoBars4h = deps.getCryptoBars4h || defaultGetCryptoBars4h;
  const placeOrder = deps.placeOrder || defaultPlaceOrder;

  const sym = toSlash(pos.symbol || "");
  const qty = Number(pos.qty || 0);
  if (qty <= 0) return null; // long-only

  try {
    const orders = await getOpenOrders(sym);
    if (orders.some((o) => o.side === "sell")) return null; // exit already in flight
  } catch {
    return null; // can't verify -> fail safe, do nothing
  }

  const psPos = ps.getPosition(state, sym);
  let entry = Number(pos.avg_entry_price || 0);
  if (entry <= 0) entry = Number(psPos.entry_price || 0);
  if (entry <= 0) return null; // no usable cost basis

  let ask, bid;
  try {
    const q = await getLatestQuote(sym);
    ask = Number(q.ap || 0);
    bid = Number(q.bp || 0);
  } catch {
    return null;
  }
  const cur = bid || ask;
  if (cur <= 0 || ask <= 0) return null;

  // 4H bars for the swing-low stop.
  let swingStop = null;
  if (STOP_LOSS_MODE === "swing_low_4h") {
    try {
      const bars4h = await getCryptoBars4h(sym);
      const lows = bars4h.filter((b) => b.c).map((b) => Number(b.l || 0));
      swingStop = swingLowStopPrice(entry, lows);
    } catch {
      // fixed -5% fallback below
    }
  }

  // Trailing stop first (supersedes the hard stop once armed).
  const hwm = psPos.high_water_mark || entry;
  const trailHit = shouldTrailStopOut(entry, hwm, cur);

  const breakeven = psPos.breakeven_stop;
  const effStop = [swingStop, breakeven].filter((s) => s).length ? Math.max(...[swingStop, breakeven].filter((s) => s)) : null;
  const hardHit = shouldStopOut(entry, cur, effStop);

  if (!trailHit && !hardHit) return null;

  const kind = trailHit
    ? "TRAILING STOP"
    : effStop && breakeven && effStop === breakeven
      ? "STOP (breakeven)"
      : effStop
        ? "STOP (4H swing low)"
        : `STOP (fixed -${Math.round(STOP_LOSS_PCT * 100)}%)`;
  const lim = stopLossLimitPrice(ask, psPos.stop_order_cycles || 0);
  const line = `${sym} ${kind}: entry $${entry.toFixed(4)} current $${cur.toFixed(4)} -> SELL ${qty.toFixed(4)} @ $${lim.toFixed(4)}`;

  if (!execute) return line + " (dry-run)";
  try {
    const result = await placeOrder(sym, qty, "sell", lim, true);
    const orderId = result.id || "";
    if (orderId) ps.setStopOrder(state, sym, orderId, lim);
    return line + " id=" + orderId.slice(0, 8);
  } catch (e) {
    if (e instanceof TradeRejected) return line + " REJECTED: " + e.message;
    return line + " ERROR: " + String(e);
  }
}

/** Build the `## Stop Watchdog HH:MM GMT+2` block text (pure -- no I/O). See journal.js's buildJournalBlockText for why this is split out. */
export function buildStopWatchdogBlockText(actions, now = new Date()) {
  const { timeStr } = amsterdamParts(now);
  const lines = ["", `## Stop Watchdog ${timeStr} GMT+2`, ""];
  for (const line of actions) lines.push("- " + line);
  return lines.join("\n") + "\n";
}

/** Append a `## Stop Watchdog HH:MM GMT+2` block (only called when something acted). */
export function appendStopWatchdogBlock(actions, now = new Date(), journalDir = JOURNAL_DIR) {
  mkdirSync(journalDir, { recursive: true });
  const { dateStr } = amsterdamParts(now);
  const filePath = path.join(journalDir, dateStr + ".md");
  appendFileSync(filePath, buildStopWatchdogBlockText(actions, now), "utf-8");
  return filePath;
}

/**
 * Run one watchdog pass. Returns a process-style exit code (0 success, 1
 * hard failure). `execute=false` (the default) is a dry run.
 */
export async function main({ execute = false, deps = {} } = {}) {
  const getPositions = deps.getPositions || defaultGetPositions;
  const loadState = deps.loadState || ps.loadState;
  const saveState = deps.saveState || ps.saveState;
  const appendBlock = deps.appendStopWatchdogBlock || appendStopWatchdogBlock;
  const now = deps.now || (() => new Date());

  const state = loadState();
  let positions;
  try {
    positions = await getPositions();
  } catch (e) {
    console.error("FAIL: positions fetch: " + e);
    return 1;
  }

  const longs = positions.filter((p) => Number(p.qty || 0) > 0);
  const actions = [];
  for (const p of longs) {
    const line = await checkPosition(p, state, execute, deps);
    if (line) {
      actions.push(line);
      console.log(line);
    }
  }

  if (!actions.length) {
    console.log(`stop watchdog: ${longs.length} long position(s) checked, no stops hit`);
    return 0;
  }

  // Journal + state only when something happened (a no-op run must not
  // generate churn).
  if (execute) saveState(state);
  const journalPath = appendBlock(actions, now());
  console.log("Wrote: " + journalPath);
  return 0;
}

// CLI entrypoint, mirrors runEvaluation.js.
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const execute = process.argv.includes("--execute");
  main({ execute }).then((code) => process.exit(code));
}
